[BUGFIX] Permettre aux étudiants faisant partie d'une orga SCO non isManagingStudents d'entrer en certif (PIX-2388)

Anne-Gaelle-S opened the pull request:
## :unicorn: Problème
Il existe 2 types d'établissements SCO : ceux important leurs élèves depuis leur fichier SIECLE (aussi appelés "isManagingStudent") et ceux n'important pas leurs élèves (et donc n'ayant pas de "gestion d'élèves").

En ce qui concerne Pix Certif, nous avons aménagé l'ajout des candidats pour les établissements SCO et "isManagingStudent" afin qu'ils puisse rajouter leurs élèves via la liste de leurs élèves déjà importés dans nos bases de données. Pour les établissements SCO et non "isManagingStudent" , les membres de Pix Certif profitent de la même interface que les centres de certification non SCO, c'est à dire avec un imports de candidats via un fichier CSV.

Pour l'application Mon-pix, lorsqu'un candidat venant du SCO tente de rejoindre une session de certification, nous faisons en sorte de vérifier si le candidat est bien connecté avec le même compte que celui avec lequel il a passé ses campagnes (passées via son établissement). Cette étape de réconciliation permet de vérifier le lien entre un candidat (certification-candidate) et un étudiant (schooling-registration).

**Un étudiant venant d'un établissement SCO et non "isManagingStudent" ne peux pas se connecter à sa certification à cause de la vérification de réconciliation.**

## :robot: Solution
Permettre aux étudiants venant des établissements "SCO" et non "isManagingStudent" de rejoindre une certification. Soit enlever la vérification de réconciliation vers un étudiant de base SIECLE pour ce genre de cas. 

## :rainbow: Remarques
Est-ce qu'on devrait déplacer la liaison entre un centre de certif et une orga dans la couche domain au lieu de laisser les repo vérifier ça ? 

## :100: Pour tester
- Lancer l'API et toutes les app
- Créer un organization SCO et non isManagingStudent (PixAdmin)
- Créer un centre de certif avec le même externalId que cette orga (PixAdmin)
- Créer un membre pour ce centre (PixAdmin)
- Créer une session (PixCertif)
- Inscrire 1 candidat de certification (PixCertif)
- Tenter de rejoindre la session avec un compte certifiable (PixApp)
- Ca devrait fonctionner (on doit pouvoir rejoindre le test de certif)

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2754.review.pix.fr
- Orga: https://orga-pr2754.review.pix.fr
- Certif: https://certif-pr2754.review.pix.fr
- Admin: https://admin-pr2754.review.pix.fr
- API: https://api-pr2754.review.pix.fr/api/

 Please check it out!

Anne-Gaelle-S committed:
[API] Remove dead code about isSco

P-Jeremy commented on api/lib/domain/models/Organization.js:
```diff
@@ -75,6 +75,10 @@ class Organization {
   }
+
+  get isScoAndManagingStudents() {
+    return this.isSco && this.isManagingStudents === true;
```
```suggestion
    return this.isSco && this.isManagingStudents;
```

Suggestion: "=== true" pas nécessaire, si je ne m'abuse, après je ne suis pas le champion des booléens... :smile:

Anne-Gaelle-S replied:
Yeppa je sais, c'était un partis pris pour être explicite et lire ça comme du lait (oui cette expression n'existe pas x))

P-Jeremy replied:
Du coup pourquoi ne pas ajouter "=== true" à this.isSco aussi ? :thinking: 

ou ```    return (this.isSco && this.isManagingStudents) === true;```

Anne-Gaelle-S replied:
Parce que c'est une propriété déjà définie (peut être qu'on voudrait pouvoir la refacto et qu'elle renvoie 'true' un jour 🤔 )

Libouk replied:
Enlever le === true

lisequesnel commented:
>Il existe 2 types d'établissements SCO : ceux important leurs élèves depuis leur fichier SIECLE (aussi appelés "isManagingStudent" et ceux important leurs élèves autrement (sans fichier SIECLE).

Alors, ce n'est pas tout à fait ça ^^
Il y a les organisations SCO qui gèrent des élèves (`isManagingStudents`) et donc fonctionnent avec un import (XML SIECLE ou CSV), et les organisations qui n'ont pas cette gestion d'élèves (`isManagingStudents` à false) et donc qui ne réalisent pas d'import du tout.

P-Jeremy commented on api/lib/domain/usecases/link-user-to-session-certification-candidate.js:
```diff
@@ -81,6 +86,18 @@ async function _getSessionCertificationCandidateByPersonalInfo({
+  organizationRepository,
+}) {
+  const { certificationCenterId } = await sessionRepository.get(sessionId);
+  const organizationBehindThisSession = await organizationRepository
```
```suggestion
  const sessionOrganization = await organizationRepository
```

:thinking:

Anne-Gaelle-S replied:
Pourquoi faire simple quand on peut faire compliqué 🤣 (non bon j'avoue j'ai écris ce que je me disais à voix haute)

Anne-Gaelle-S replied:
Bon ceci dit je me pose vraiment la question de est-ce que le terme "sessionOrganization" est assez explicite 🤔 . Est-ce qu'on peut parler de l'organisation d'une session tout simplement ?

P-Jeremy replied:
Ma suggestion n'est effectivement peut être pas assez explicite... J'ai pas mieux, après c'est du chipotage, rien de bloquant :wink:

Anne-Gaelle-S replied:
J'aime bien ton chipotage, j'ai d'ailleurs simplifié avec ta suggestion. Je laisse quand même le débat ouvert 😈

laura-bergoens commented on api/lib/domain/models/Organization.js:
```diff
@@ -75,6 +75,10 @@ class Organization {
     return Boolean(this.tags.find((tag) => this.isPro && tag.name === Tag.MEDIATION_NUMERIQUE));
   }
+
+  get isScoAndManagingStudents() {
```
Ayant été un petit moment en team prescription, ils sont familiers avec le terme isScoManagingStudents(), tu peux enlever le "and" je pense

Anne-Gaelle-S replied:
Ne connaissant pas aussi bien le périmètre que la team prescription, je trouve ça moins explicite sans le and 🤔 ... si c'est le cas d'autres personnes hors prescription, peut être que ça vaut le coup de le laisser

Libouk replied:
Quand on l'avait mis en place on avait discuté du fait que ça pouvait être deux notions distinctes capables de "vivre" l'une sans l'autre (Sco // Managing student) et que du coup ça avait du sens de mettre le And

Anne-Gaelle-S committed:
[API] Add isScoAndManagingStudents in orga model

Anne-Gaelle-S committed:
[API] Add getBySessionId in certifcenter repo

Anne-Gaelle-S committed:
[API] Add getScoOrganizationByExternalId() repo

Anne-Gaelle-S committed:
[API][SR] Remove dupplicated test

Anne-Gaelle-S committed:
[API][SR] Remove already existing tests

Theses tests should not be in this unit usecase test.
Futhermore, there are already define in CertificationCandidate_test.js.

Anne-Gaelle-S committed:
[API][SR] Remove beforeEach in test

Anne-Gaelle-S committed:
[API][SR] Remove useless configuration

Anne-Gaelle-S committed:
[API] Change SCO check in certif joiner usecase

There are two type of SCO organization.
For all candidates from SCO orga with isManagingStudents to false,
we should apply same validation as candidates from PRO or SUP.

Libouk commented on api/tests/unit/domain/models/Organization_test.js:
```diff
@@ -300,4 +300,32 @@ describe('Unit | Domain | Models | Organization', () => {
+      expect(organization.isScoAndManagingStudents).is.false;
+    });
+
+    it('should return false when organization is not managingStudent', () => {
```
Commentaire concernant les deux cas SCO : Est-ce que ce serait mieux de les réunir dans un context SCO avec deux It pour managingStudent true/false ?

Anne-Gaelle-S replied:
Hmmm pour ce `it` là on pourrait mettre autre chose que du `SCO`. L'important est que isManagingStudent soit à `false`. Du coup, je pense que les mettre dans un context "sco" peut apporter une info pas forcément nécessaire ici 🤔

Libouk commented on api/lib/infrastructure/repositories/organization-repository.js:
```diff
@@ -100,6 +100,19 @@ module.exports = {
+  async getScoOrganizationByExternalId(externalId) {
+    return BookshelfOrganization
+      .query((qb) => qb.where({ type: Organization.types.SCO })
+        .whereRaw('LOWER("externalId") = ? ', `${externalId.toLowerCase()}`))
```
externalId.toLowerCase (sans les `)

Libouk commented on api/tests/integration/infrastructure/repositories/organization-repository_test.js:
```diff
@@ -251,6 +251,55 @@ describe('Integration | Repository | Organization', function() {
 
+  describe('#getScoOrganizationByExternalId', () => {
+
+    describe('when there is a organization with given externalId', () => {
```
aN organization

Libouk commented on api/tests/integration/infrastructure/repositories/organization-repository_test.js:
```diff
@@ -251,6 +251,55 @@ describe('Integration | Repository | Organization', function() {
+        expect(result.id).to.deep.equal(1);
+        expect(result.type).to.deep.equal('SCO');
+        expect(result.externalId).to.deep.equal('1234567');
+        expect(result.isManagingStudents).to.deep.equal(true);
```
Vérifier que ce soit une instance d'une organisation

Libouk commented on api/tests/integration/infrastructure/repositories/organization-repository_test.js:
```diff
@@ -251,6 +251,55 @@ describe('Integration | Repository | Organization', function() {
+
+    describe('when there is no organization with given externalId', () => {
+
+      it('should throw an error if the id does not match an certification center with organization ', async () => {
```
id -> externalId, an -> a (Voilà le N volé plus haut)

Libouk commented on api/tests/unit/domain/usecases/link-user-to-session-certification-candidate_test.js:
```diff
@@ -301,96 +262,229 @@ describe('Unit | Domain | Use Cases | link-user-to-session-certification-candida
-            sessionRepository,
+    context('when the organization is not managing students', () => {
+
+      it('should create a link with between the candidate and the user and return the linked certification candidate', async () => {
```
Enlever le with ? Jerem se demande si "should return the linked certification candidate" ne serait pas suffisant ?

laura-bergoens commented on api/lib/infrastructure/repositories/certification-center-repository.js:
```diff
@@ -46,6 +46,21 @@ module.exports = {
       });
   },
 
+  getBySessionId(sessionId) {
```
Je pense que tu devrais toujours faire du async/await, c'est plus joli. Je me doute que tu as voulu garder l'homogénéité dans le fichier, mais osef, go faire du code moderne ! On pourrait d'ailleurs imaginer un commit de BSR d'async/await des autres méthodes (le fichier n'est pas très gros, c'est rapide)

laura-bergoens commented on api/tests/integration/infrastructure/repositories/certification-center-repository_test.js:
```diff
@@ -46,6 +46,52 @@ describe('Integration | Repository | Certification Center', () => {
+        await databaseBuilder.commit();
+
+        // when
+        const promise = certificationCenterRepository.getBySessionId(8);
+
+        // then
+        return expect(promise).to.have.been.rejectedWith(NotFoundError);
```
```suggestion
        const error = await catchErr(certificationCenterRepository.getBySessionId)(8);

        // then
        expect(error).to.be.instanceOf(NotFoundError);
```

laura-bergoens commented on api/lib/infrastructure/repositories/organization-repository.js:
```diff
@@ -100,6 +100,19 @@ module.exports = {
     throw new NotFoundError(`Not found organization for certification center id ${certificationCenterId}`);
   },
 
+  async getScoOrganizationByExternalId(externalId) {
```
idem, tu déclares la méthode async mais tu ne te sers pas d'await. Utilise await c'est moderne

laura-bergoens commented on api/lib/infrastructure/repositories/organization-repository.js:
```diff
@@ -100,6 +100,19 @@ module.exports = {
+  async getScoOrganizationByExternalId(externalId) {
+    return BookshelfOrganization
+      .query((qb) => qb.where({ type: Organization.types.SCO })
+        .whereRaw('LOWER("externalId") = ? ', externalId.toLowerCase()))
```
```suggestion
        .whereRaw('LOWER("externalId") = ?', externalId.toLowerCase()))
```
#casse_bonbons

laura-bergoens commented on api/lib/infrastructure/repositories/certification-center-repository.js:
```diff
@@ -46,6 +46,21 @@ module.exports = {
+      })
+      .fetch()
+      .then((certificationCenter) => {
+        if (certificationCenter) {
+          return _toDomain(certificationCenter);
+        }
```
D'après moi, tu ne devrais jamais rentrer dans ce if (tout comme l'autre plus loin), mais j'ai beau tester chez moi ça marche pas comme ça devrait marcher, on regarde ça tout à l'heure

laura-bergoens replied:
la réponse à ce comportement bizarre est ici -> api/lib/infrastructure/data/certification-center.js:9

Anne-Gaelle-S committed:
[API][SR] Async/await certif center repo methods
